File: scripts/changelog.py
from __future__ import annotations

import argparse
from pathlib import Path
import re
import sys
from typing import Any

from scripts.common import ROOT, ToolError, error_json, run_bounded, write_new
from scripts.release import checked_revision, checked_version, workspace_version

GROUPS = ('Features', 'Fixes', 'Performance', 'Tests', 'Documentation', 'Build and maintenance', 'Other')
KINDS = {'feat': 'Features', 'fix': 'Fixes', 'perf': 'Performance', 'test': 'Tests',
         'docs': 'Documentation', 'build': 'Build and maintenance', 'ci': 'Build and maintenance',
         'chore': 'Build and maintenance', 'refactor': 'Build and maintenance'}


def render_notes(records: list[tuple[str, str]], version: str, revision: str) -> str:
    checked_version(version)
    checked_revision(revision)
    groups: dict[str, list[str]] = {group: [] for group in GROUPS}
    for sha, subject in records:
        checked_revision(sha)
        if not subject or '\n' in subject or '\r' in subject:
            raise ToolError('INVALID_COMMIT_SUBJECT', 'Commit subjects must be nonempty single lines')
        match = re.match(r'^([a-z]+)(?:\([^)]*\))?!?:\s+', subject)
        group = KINDS.get(match.group(1), 'Other') if match else 'Other'
        escaped = re.sub(r'([\\`*{}\[\]()<>_#])', r'\\\1', subject)
        groups[group].append(f'- {escaped} ({sha[:12]})')
    lines = [f'# DocSight {version}', '', f'Candidate commit: `{revision}`.', '',
             'Generated from non-merge Git commits. Publishing requires the release validation gates.', '']
    for group, entries in groups.items():
        if entries:
            lines += [f'## {group}', '', *entries, '']
    if not records:
        lines += ['No non-merge commits exist in the requested range.', '']
    return '\n'.join(lines)


def generate_notes(root: Path, version: str, revision: str, since: str | None = None) -> str:
    checked_revision(revision)
    if since is not None:
        checked_revision(since)
        ancestry = run_bounded(['git', 'merge-base', '--is-ancestor', since, revision], cwd=root)
        if ancestry.returncode != 0 or ancestry.termination:
            raise ToolError('INVALID_CHANGELOG_RANGE', 'Previous revision must be an ancestor of the candidate')
    selection = f'{since}..{revision}' if since is not None else revision
    result = run_bounded(['git', 'log', '--reverse', '--no-merges', '--format=%H%x00%s', selection, '--'],
                         cwd=root, output_limit=2_097_152)
    if result.returncode != 0 or result.termination:
        raise ToolError('GIT_HISTORY_UNAVAILABLE', 'Cannot read the requested bounded Git history')
    records = []
    for line in result.stdout.decode('utf-8').splitlines():
        fields = line.split('\0', 1)
        if len(fields) != 2:
            raise ToolError('INVALID_GIT_OUTPUT', 'Git history output does not match the requested format')
        records.append((fields[0], fields[1]))
    return render_notes(records, version, revision)


def main() -> int:
    parser = argparse.ArgumentParser(description='Generate release notes from verified local Git history')
    parser.add_argument('--revision', required=True)
    parser.add_argument('--since')
    parser.add_argument('--out', type=Path, required=True)
    arguments = parser.parse_args()
    try:
        output = generate_notes(ROOT, workspace_version(), arguments.revision, arguments.since)
        write_new(arguments.out, output.encode(), 0o644)
        return 0
    except (ToolError, OSError, ValueError) as error:
        sys.stderr.buffer.write(error_json(error))
        return 2


if __name__ == '__main__':
    raise SystemExit(main())

File: xtask/src/release/changelog.rs
use crate::tooling::{
    common::*,
    process::{ProcessLimits, run_bounded},
};
use std::path::Path;

pub const GROUPS: [&str; 7] = [
    "Features",
    "Fixes",
    "Performance",
    "Tests",
    "Documentation",
    "Build and maintenance",
    "Other",
];

fn group(subject: &str) -> usize {
    let Some((prefix, body)) = subject.split_once(':') else {
        return 6;
    };
    if !body.starts_with(char::is_whitespace) {
        return 6;
    }
    let prefix = prefix.trim_end_matches('!');
    let kind = if let Some((kind, scope)) = prefix.split_once('(') {
        if !scope.ends_with(')') || scope[..scope.len() - 1].contains(')') {
            return 6;
        }
        kind
    } else {
        prefix
    };
    match kind {
        "feat" => 0,
        "fix" => 1,
        "perf" => 2,
        "test" => 3,
        "docs" => 4,
        "build" | "ci" | "chore" | "refactor" => 5,
        _ => 6,
    }
}

pub fn render(records: &[(String, String)], version: &str, revision: &str) -> Result<String> {
    checked_version(version)?;
    checked_revision(revision)?;
    let mut grouped: [Vec<String>; 7] = std::array::from_fn(|_| Vec::new());
    for (sha, subject) in records {
        checked_revision(sha)?;
        require(
            !subject.is_empty() && subject.len() <= 8192 && !subject.contains(['\n', '\r', '\0']),
            "INVALID_COMMIT_SUBJECT",
            "Commit subjects must be bounded nonempty single lines",
        )?;
        let mut escaped = String::new();
        for character in subject.chars() {
            if "\\`*{}[]()<>_#".contains(character) {
                escaped.push('\\');
            }
            escaped.push(character);
        }
        grouped[group(subject)].push(format!("- {escaped} ({})", &sha[..12]));
    }
    let mut output = format!(
        "# DocSight {version}\n\nCandidate commit: `{revision}`.\n\nGenerated from non-merge Git commits. Publishing requires the release validation gates.\n\n"
    );
    for (name, entries) in GROUPS.into_iter().zip(grouped) {
        if !entries.is_empty() {
            output.push_str(&format!("## {name}\n\n{}\n\n", entries.join("\n")));
        }
    }
    if records.is_empty() {
        output.push_str("No non-merge commits exist in the requested range.\n");
    } else {
        output.pop();
    }
    Ok(output)
}

pub fn generate(root: &Path, version: &str, revision: &str, since: Option<&str>) -> Result<String> {
    checked_revision(revision)?;
    if let Some(since) = since {
        checked_revision(since)?;
        let result = run_bounded(
            &["git", "merge-base", "--is-ancestor", since, revision],
            root,
            &ProcessLimits::default(),
            None,
        )?;
        require(
            result.returncode == 0 && result.termination.is_none(),
            "INVALID_CHANGELOG_RANGE",
            "Previous revision must be an ancestor of the candidate",
        )?;
    }
    let selection = since
        .map(|since| format!("{since}..{revision}"))
        .unwrap_or_else(|| revision.to_owned());
    let result = run_bounded(
        &[
            "git",
            "log",
            "--reverse",
            "--no-merges",
            "--format=%H%x00%s",
            &selection,
            "--",
        ],
        root,
        &ProcessLimits {
            output_bytes: 2_097_152,
            ..ProcessLimits::default()
        },
        None,
    )?;
    require(
        result.returncode == 0 && result.termination.is_none(),
        "GIT_HISTORY_UNAVAILABLE",
        "Cannot read requested bounded Git history",
    )?;
    let mut records = Vec::new();
    for line in text(&result.stdout)?.lines() {
        let (sha, subject) = line.split_once('\0').ok_or_else(|| {
            ToolError::new(
                "INVALID_GIT_OUTPUT",
                "Git history does not match its requested format",
            )
        })?;
        records.push((sha.to_owned(), subject.to_owned()));
    }
    render(&records, version, revision)
}
